### PR TITLE
add: calculate reward by liquidity share

### PR DESCRIPTION
**Pull request by ruibinx**
{Summary}

Fixes #{issue}

**Type of change**

* [ ] Bug Fix (non-breaking change which fixes an issue)
* [ ] New Feature (non-breaking change which adds functionality)
* [ ] Breaking Change (fix or feature that would cause existing functionality to not work as expected)
* [ ] This change requires a documentation update
* [ ] This change requires a storage migration
* [ ] This change requires a benchmark

**Checklist**

* [ ] My code follows the style guidelines of this project
* [ ] I have performed a self-review of my own code
* [ ] I have commented my code, particularly in hard-to-understand areas
* [ ] I have made corresponding changes to the documentation
* [ ] My changes generate no new warnings
* [ ] I have added tests that prove my fix is effective or that my feature works
* [ ] New and existing unit tests pass locally with my changes
* [ ] New and existing benchmarks pass locally with my changes
* [ ] Any dependent changes have been merged and published in downstream modules
* [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: pallets/swap/src/impl_swaps.rs [hunk 1]
```diff
@@ -196,7 +198,16 @@ impl<T: Config> Swaps<AccountOf<T>> for Pallet<T> {
             <Provider<T>>::insert(liquidity.token_id, &who, holding);
         }
 
+        let current_block = <frame_system::Pallet<T>>::block_number();
+
+        meta.liquidity_share
```
Comment (coderfengyun) [inline pallets/swap/src/impl_swaps.rs]: addLiquidity时也需要做同样的事情

File: pallets/swap/src/functions.rs [hunk 2]
```diff
@@ -176,4 +179,16 @@ impl<T: Config> Pallet<T> {
 
         Ok(result)
     }
+
+    pub(super) fn calculate_liquidity_share(
```
Comment (coderfengyun) [inline pallets/swap/src/functions.rs]:
```suggestion
    pub(super) fn calculate_liquidity_contribution(
```

File: pallets/swap/src/impl_swaps.rs [hunk 3]
```diff
@@ -139,6 +141,12 @@ impl<T: Config> Swaps<AccountOf<T>> for Pallet<T> {
             holding.saturating_accrue(liquidity);
         });
 
+        meta.liquidity_share
+            .saturating_accrue(Pallet::<T>::calculate_liquidity_share(
+                minted - meta.updated_at,
```
Comment (coderfengyun) [inline pallets/swap/src/impl_swaps.rs]: swap初始化的同时会mint，此时updated_at == 0, minted > 0, 但是这个计算结果是错的。
mint这里应该是minted - max(meta.create_at, meta.updated_at)

Comment (coderfengyun) [reply pallets/swap/src/impl_swaps.rs]: 增加对应测试用例

File: pallets/swap/src/tests.rs [hunk 4]
```diff
@@ -439,3 +439,128 @@ fn should_not_overflow_when_calculating_price() {
...
+            let (_, token) = Swap::calculate_reward(lp_token_id).unwrap();
+            actual_token += token;
+            println!("token {}, {}", lp_token_id, token);
+        }
+
+        let total_liquidity_share = 5 * 10 * 10 + 100 * 10;
+
+        assert_eq!(
+            actual_token,
+            10 * 20 * total_reward / total_liquidity_share * 5
```
Comment (coderfengyun) [inline pallets/swap/src/tests.rs]: 10 * 20 * 5是什么含义?

File: pallets/swap/src/impl_swaps.rs (hunk 1 above)
Comment (ruibinx) [reply pallets/swap/src/impl_swaps.rs]: done

File: pallets/swap/src/impl_swaps.rs (hunk 3 above)
Comment (ruibinx) [reply pallets/swap/src/impl_swaps.rs]: 见53行，updated初始化时等于created_at

File: pallets/swap/src/tests.rs (hunk 4 above)
Comment (ruibinx) [reply pallets/swap/src/tests.rs]: 10 liquidity * 20 blocks for 5 lp tokens

**Review by coderfengyun — APPROVED** [review]

**Pull request closed without merging**